File: src/main/java/com/sidgate/poker/Poker.java
package com.sidgate.poker;

import com.sidgate.deck.Hand;

public class Poker {
	private final HandType bestHandTypeChain;

	public Poker() {
		bestHandTypeChain = new OfAKind(4, "four-of-a-kind",
			new FullHouse(
				new Flush(
					new Straight(
						new OfAKind(3, "three-of-a-kind",
							new OfAKind(2, "pair",
								new HighCard()))))));
	}

	public String bestHand(Hand hand) {
		return bestHandTypeChain.getBestHand(hand).getName();
	}
}

File: src/main/kotlin/sidgate/poker/Poker.kt
package sidgate.poker

import sidgate.deck.Hand

class Poker {
    private val bestHandTypeChain: HandType
    fun bestHand(hand: Hand): String {
        return bestHandTypeChain.getBestHand(hand).name
    }

    init {
        bestHandTypeChain = OfAKind(4, "four-of-a-kind",
                FullHouse(
                        Flush(
                                Straight(
                                        OfAKind(3, "three-of-a-kind",
                                                OfAKind(2, "pair",
                                                        HighCard()))))))
    }
}